traffic_router/core/src/main/java/org/apache/traffic_control/traffic_router/core/http/BufferedResponseFilter.java

```diff
@@ -0,0 +1,46 @@
+/*
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.traffic_control.traffic_router.core.http;
+
+import com.google.common.net.HttpHeaders;
+import org.apache.log4j.Logger;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class BufferedResponseFilter extends OncePerRequestFilter {
+	public static final Logger LOGGER = Logger.getLogger(BufferedResponseFilter.class);
+
+	public void doFilterInternal(final HttpServletRequest request, final HttpServletResponse response, final FilterChain chain) throws IOException, ServletException {
+		final BufferedResponse responseWrapper = new BufferedResponse(response);
+
+		chain.doFilter(request, responseWrapper);
+
+		// Close the connection without waiting for the 10-second connect timeout,
+		// in case the client does not close the connection. Even though this is
+		// the only case for which we are interested in sending Connection: close,
+		// sending it sometimes means we must always send it. From RFC 2616:
+		// > HTTP/1.1 applications that do not support persistent connections MUST
+		// > include the "close" connection option in every message.
+		response.addHeader(HttpHeaders.CONNECTION, "close");
```

Comment:
This worries me, because we might be fixing the performance of _invalid_ requests (e.g. `curl -X HEAD`), at the expense of _valid_ requests. Keeping client connections alive where possible can be a good thing.

Reply:
Does TR support HTTP/2? The spec for HTTP/2 says that clients must treat any message containing a connection-specific header as malformed.

Reply:
`close` is the default behavior for HTTP/1.0, so really you just want to add this header if the request is HTTP/1.1. You can check that with the request's `getProtocol` method, I think.

Reply:
> Does TR support HTTP/2?

No, and `curl -X HEAD` doesn't hang for HTTP/2 anyway.

As an alternative to `Connection: close`, we could just reduce `Connector.connectionTimeout` from 10 seconds to something shorter, like 3 or 5 seconds:

https://github.com/apache/trafficcontrol/blob/b24649869407d902301ce6997c629e367285b44f/traffic_router/core/src/main/conf/server.xml#L38

https://github.com/apache/trafficcontrol/blob/b24649869407d902301ce6997c629e367285b44f/traffic_router/core/src/main/conf/server.xml#L41

https://github.com/apache/trafficcontrol/blob/b24649869407d902301ce6997c629e367285b44f/traffic_router/core/src/main/conf/server.xml#L43

https://github.com/apache/trafficcontrol/blob/b24649869407d902301ce6997c629e367285b44f/traffic_router/core/src/main/conf/server.xml#L47

Reply:
Ideally @rawlinp would chime in to say whether or not that's acceptable

Reply:
`Connection: close` still kind of worries me, especially because there may be clients out there relying on high connection rates to TR, whether that is really the _right_ approach or not, reusing connections could actually be helping TR performance. Maybe I don't fully understand the need for it here? Are `HEAD` requests broken without it?

Changing `connectionTimeout` in server.xml seems like it would be more fine-tuned via the CDN operator's ansible playbooks (or equivalent) if the CDN had issues with slow clients. I think Tomcat ships with a 20s default, and 10s seems reasonable for a TR default.

Reply:
> Looks like cloudfront.com is trying to close the connection but mistyped it as `Cneonction`:

As @rob05c pointed out to me, `Cneonction` is just [a hack](https://stackoverflow.com/questions/4798461/cneonction-and-nncoection-http-headers) to get around Connection header differences between the origin and caches.

> Maybe I don't fully understand the need for it here? Are `HEAD` requests broken without it?

Nope we don't need it, it would only be to appease invalid requests like `curl -X HEAD`. But `curl -X HEAD` doesn't cause a timeout anyway in HTTP/2, so there isn't a long-term reason to do it, especially with the benefits from reusing the connection like you mentioned.

Even if we did want to change any connection behavior, that belongs in a separate PR, #4074 should just be to resolve disparities between `HEAD` and `GET`.

Reply:
Forgot to mention: Reverted closing the connection in 97d38f2234